File: backend/app/api/v1/reports.py
from datetime import date
from typing import Optional
from fastapi import APIRouter, Depends, Query, Response
from app.core.deps import require_admin
from app.services.report_service import report_service

router = APIRouter(prefix="/admin/reports", tags=["Laporan PDF & Excel (Admin)"])


# 1. TRANSAKSI
@router.get("/transactions.xlsx")
def get_transactions_report_excel(
    start_date: Optional[str] = Query(None),
    end_date: Optional[str] = Query(None),
    type: Optional[str] = Query(None),
    category_id: Optional[str] = Query(None),
    nasabah_id: Optional[str] = Query(None),
    admin_user: dict = Depends(require_admin)
):
    """Download laporan transaksi format Excel (.xlsx)."""
    excel_bytes = report_service.export_transactions_excel(
        start_date=start_date, end_date=end_date, type_filter=type,
        category_id=category_id, nasabah_id=nasabah_id
    )
    filename = f"laporan-transaksi-{start_date or 'all'}-{end_date or date.today().isoformat()}.xlsx"
    return Response(
        content=excel_bytes,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f"attachment; filename={filename}"}
    )


@router.get("/transactions.pdf")
def get_transactions_report_pdf(
    start_date: Optional[str] = Query(None),
    end_date: Optional[str] = Query(None),
    type: Optional[str] = Query(None),
    category_id: Optional[str] = Query(None),
    nasabah_id: Optional[str] = Query(None),
    admin_user: dict = Depends(require_admin)
):
    """Download laporan transaksi format PDF (.pdf)."""
    pdf_bytes = report_service.export_transactions_pdf(
        start_date=start_date, end_date=end_date, type_filter=type,
        category_id=category_id, nasabah_id=nasabah_id
    )
    filename = f"laporan-transaksi-{start_date or 'all'}-{end_date or date.today().isoformat()}.pdf"
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": f"inline; filename={filename}"}
    )


# 2. REKAP KATEGORI
@router.get("/category-recap.xlsx")
def get_category_recap_excel(
    start_date: Optional[str] = Query(None),
    end_date: Optional[str] = Query(None),
    admin_user: dict = Depends(require_admin)
):
    """Download laporan rekapitulasi setoran per kategori sampah (.xlsx)."""
    excel_bytes = report_service.export_category_recap_excel(start_date=start_date, end_date=end_date)
    filename = f"laporan-rekap-kategori-{start_date or 'all'}-{end_date or date.today().isoformat()}.xlsx"
    return Response(
        content=excel_bytes,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f"attachment; filename={filename}"}
    )


@router.get("/category-recap.pdf")
def get_category_recap_pdf(
    start_date: Optional[str] = Query(None),
    end_date: Optional[str] = Query(None),
    admin_user: dict = Depends(require_admin)
):
    """Download laporan rekapitulasi setoran per kategori sampah (.pdf)."""
    pdf_bytes = report_service.export_category_recap_pdf(start_date=start_date, end_date=end_date)
    filename = f"laporan-rekap-kategori-{start_date or 'all'}-{end_date or date.today().isoformat()}.pdf"
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": f"inline; filename={filename}"}
    )


# 3. NASABAH & SALDO
@router.get("/nasabah.xlsx")
def get_nasabah_report_excel(admin_user: dict = Depends(require_admin)):
    """Download daftar seluruh nasabah dan saldo tabungan (.xlsx)."""
    excel_bytes = report_service.export_nasabah_excel()
    filename = f"laporan-nasabah-saldo-{date.today().isoformat()}.xlsx"
    return Response(
        content=excel_bytes,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f"attachment; filename={filename}"}
    )


@router.get("/nasabah.pdf")
def get_nasabah_report_pdf(admin_user: dict = Depends(require_admin)):
    """Download daftar seluruh nasabah dan saldo tabungan (.pdf)."""
    pdf_bytes = report_service.export_nasabah_pdf()
    filename = f"laporan-nasabah-saldo-{date.today().isoformat()}.pdf"
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": f"inline; filename={filename}"}
    )


# 4. DAFTAR HARGA
@router.get("/prices.xlsx")
def get_prices_report_excel(admin_user: dict = Depends(require_admin)):
    """Download daftar master harga sampah (.xlsx)."""
    excel_bytes = report_service.export_prices_excel()
    filename = f"laporan-master-harga-{date.today().isoformat()}.xlsx"
    return Response(
        content=excel_bytes,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f"attachment; filename={filename}"}
    )


@router.get("/prices.pdf")
def get_prices_report_pdf(admin_user: dict = Depends(require_admin)):
    """Download daftar master harga sampah (.pdf)."""
    pdf_bytes = report_service.export_prices_pdf()
    filename = f"laporan-master-harga-{date.today().isoformat()}.pdf"
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": f"inline; filename={filename}"}
    )

File: backend/src/routes/reports.ts
import { Hono } from 'hono';
import { AppEnv } from '../types/index.js';
import { reportService } from '../services/reportService.js';
import { authMiddleware } from '../middleware/auth.js';
import { requireAdmin } from '../middleware/roleGuard.js';
import { rateLimiter } from '../middleware/rateLimiter.js';

export const reportsRouter = new Hono<AppEnv>();

const reportsLimiter = rateLimiter({
  windowMs: 60 * 1000,
  max: 30,
  message: 'Terlalu banyak permintaan pembuatan dokumen laporan. Silakan tunggu beberapa saat.',
});

// All routes require ADMIN and Rate Limiting
reportsRouter.use('*', authMiddleware, requireAdmin, reportsLimiter);

// 1. Transactions Report
reportsRouter.get('/transactions.xlsx', async (c) => {
  const query = c.req.query();
  const buffer = await reportService.generateTransactionsExcel({
    start_date: query.start_date,
    end_date: query.end_date,
    type: query.type,
    category_id: query.category_id,
    nasabah_id: query.nasabah_id,
  });

  c.header(
    'Content-Type',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
  );
  c.header('Content-Disposition', 'attachment; filename="laporan-transaksi.xlsx"');
  return c.body(new Uint8Array(buffer));
});

reportsRouter.get('/transactions.pdf', async (c) => {
  const query = c.req.query();
  const buffer = await reportService.generateTransactionsPdf({
    start_date: query.start_date,
    end_date: query.end_date,
    type: query.type,
    category_id: query.category_id,
    nasabah_id: query.nasabah_id,
  });

  c.header('Content-Type', 'application/pdf');
  c.header('Content-Disposition', 'inline; filename="laporan-transaksi.pdf"');
  return c.body(new Uint8Array(buffer));
});

// 2. Category Recap Report
reportsRouter.get('/category-recap.xlsx', async (c) => {
  const query = c.req.query();
  const buffer = await reportService.generateCategoryRecapExcel({
    start_date: query.start_date,
    end_date: query.end_date,
  });

  c.header(
    'Content-Type',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
  );
  c.header('Content-Disposition', 'attachment; filename="rekapitulasi-sampah.xlsx"');
  return c.body(new Uint8Array(buffer));
});

reportsRouter.get('/category-recap.pdf', async (c) => {
  const query = c.req.query();
  const buffer = await reportService.generateCategoryRecapPdf({
    start_date: query.start_date,
    end_date: query.end_date,
  });

  c.header('Content-Type', 'application/pdf');
  c.header('Content-Disposition', 'inline; filename="rekapitulasi-sampah.pdf"');
  return c.body(new Uint8Array(buffer));
});

// 3. Nasabah & Saldo Report
reportsRouter.get('/nasabah.xlsx', async (c) => {
  const buffer = await reportService.generateNasabahExcel();
  c.header(
    'Content-Type',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
  );
  c.header('Content-Disposition', 'attachment; filename="daftar-nasabah-saldo.xlsx"');
  return c.body(new Uint8Array(buffer));
});

reportsRouter.get('/nasabah.pdf', async (c) => {
  const buffer = await reportService.generateNasabahPdf();
  c.header('Content-Type', 'application/pdf');
  c.header('Content-Disposition', 'inline; filename="daftar-nasabah-saldo.pdf"');
  return c.body(new Uint8Array(buffer));
});

// 4. Prices Report
reportsRouter.get('/prices.xlsx', async (c) => {
  const buffer = await reportService.generatePricesExcel();
  c.header(
    'Content-Type',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
  );
  c.header('Content-Disposition', 'attachment; filename="master-harga-sampah.xlsx"');
  return c.body(new Uint8Array(buffer));
});

reportsRouter.get('/prices.pdf', async (c) => {
  const buffer = await reportService.generatePricesPdf();
  c.header('Content-Type', 'application/pdf');
  c.header('Content-Disposition', 'inline; filename="master-harga-sampah.pdf"');
  return c.body(new Uint8Array(buffer));
});
